File: bindings/node/src/zlink/contracts/eventing/poller.ts
// SPDX-License-Identifier: MPL-2.0

import type { BaseSocket } from '../sockets';
import type { PollEventFlagValue } from '../sockets/socket_constants';
import type { SocketMonitor } from './monitor';
import type { Timer } from './timer';

/** A source that can be identified in a poll result; numbers are raw file descriptors. */
export type Pollable = BaseSocket | SocketMonitor | Timer | number;

/** One ready source reported by {@link Poller.wait}. */
export interface PollEvent {
  /** The registered object, or the raw file descriptor, that became ready. */
  readonly source: Pollable;
  /** Whether the ready source is a socket, file descriptor, or timer. */
  readonly sourceKind: number;
  /** The caller token supplied when the source was registered. */
  readonly slot: number;
  /** The events that fired, as a mask of poll flags. */
  readonly revents: number;
  /** The file descriptor, when the source is a raw fd. */
  readonly fd: number;
}

/** A reusable buffer of poll results filled by {@link Poller.wait}. */
export interface PollEvents {
  /** The maximum number of results the buffer can hold. */
  readonly capacity: number;
  /** The number of ready sources written by the last wait. */
  readonly readyCount: number;
  /** The source kind of the result at `index`. */
  sourceKind(index: number): number;
  /** The registered object, or raw file descriptor, for the result at `index`. */
  source(index: number): Pollable;
  /** The caller token of the result at `index`. */
  slot(index: number): number;
  /** The fired-event mask of the result at `index`. */
  revents(index: number): number;
  /** The file descriptor of the result at `index`. */
  fd(index: number): number;
  /** Whether the result at `index` includes `event`. */
  hasEvent(index: number, event: PollEventFlagValue): boolean;
  /** Release the buffer. */
  close(): void;
}

/** Multiplexes sockets, socket monitors, file descriptors, and timers. */
export interface Poller {
  /** The number of registered sources. */
  readonly size: number;
  /** Register `socket` to be watched for `events`; `slot` is echoed back in the matching result. */
  add(socket: BaseSocket, events: readonly PollEventFlagValue[], slot: number): void;
  /** Register `monitor`; socket monitors support `PollIn` only. */
  add(monitor: SocketMonitor, events: readonly PollEventFlagValue[], slot: number): void;
  /** Register `timer`; its expirations surface as poll events tagged with `slot`. */
  add(timer: Timer, slot: number): void;
  /** Change the watched events for an already-registered socket. */
  modify(socket: BaseSocket, events: readonly PollEventFlagValue[]): void;
  /** Change a monitor registration; socket monitors support `PollIn` only. */
  modify(monitor: SocketMonitor, events: readonly PollEventFlagValue[]): void;
  /** Unregister `socket`; return true when it was registered. */
  remove(socket: BaseSocket): boolean;
  /** Unregister `monitor`; return true when it was registered. */
  remove(monitor: SocketMonitor): boolean;
  /** Unregister `timer`; return true when it was registered. */
  remove(timer: Timer): boolean;
  /** Register raw file descriptor `fd` to be watched for `events`; `slot` is echoed back. */
  addFd(fd: number, events: readonly PollEventFlagValue[], slot: number): void;
  /** Change the watched events for an already-registered file descriptor. */
  modifyFd(fd: number, events: readonly PollEventFlagValue[]): void;
  /** Unregister file descriptor `fd`; return true when it was registered. */
  removeFd(fd: number): boolean;
  /**
   * Wait up to `timeoutMs` for sources to become ready, filling `events`; a
   * negative timeout blocks indefinitely. `PollOut` is the WRITABLE wake for
   * managed sends; `PollCompletion` means the owned completion queue
   * (REQUEST and any WRITABLE records) was drained.
   */
  wait(events: PollEvents, timeoutMs: number): number;
  /** Close the poller and release its resources. */
  close(): void;
}

File: bindings/python/src/zlink/contracts/eventing/poller.py
# SPDX-License-Identifier: MPL-2.0

from dataclasses import dataclass
from typing import Protocol, runtime_checkable

from .codes import PollEventFlag, PollSourceKind
from .monitor import MonitorSocket


@dataclass(frozen=True)
class PollEvent:
    """One ready source reported by :meth:`Poller.wait`.

    Attributes:
        source_kind: Whether the ready source is a socket, file descriptor, or timer.
        slot: The caller token supplied when the source was registered.
        revents: The events that fired, as a mask of poll flags.
        fd: The file descriptor, when the source is a raw fd.
    """

    source_kind: PollSourceKind
    slot: int
    revents: int
    fd: int = 0


@runtime_checkable
class PollEvents(Protocol):
    """A reusable buffer of poll results filled by :meth:`Poller.wait`."""

    @property
    def capacity(self):
        """The maximum number of results the buffer can hold."""
        ...

    @property
    def ready_count(self):
        """The number of ready sources written by the last wait."""
        ...

    def source_kind(self, index):
        """Return the source kind of the result at ``index``."""
        ...

    def slot(self, index):
        """Return the caller token of the result at ``index``."""
        ...

    def revents(self, index):
        """Return the fired-event mask of the result at ``index``."""
        ...

    def has_event(self, index, event):
        """Return whether the result at ``index`` includes ``event``."""
        ...

    def fd(self, index):
        """Return the file descriptor of the result at ``index``."""
        ...

    def event(self, index):
        """Return the result at ``index`` as a :class:`PollEvent`."""
        ...


@runtime_checkable
class Poller(Protocol):
    """Multiplexes sockets, socket monitors, file descriptors, and timers,
    reporting which become ready on a single wait."""

    def add_socket(self, socket, events, slot):
        """Register ``socket`` to be watched for ``events``; ``slot`` is echoed
        back in the matching result. Also accepts :class:`MonitorSocket` with
        ``POLLIN`` only; drain it with ``recv(flags=RecvFlags.DONT_WAIT)``."""
        ...

    def add_monitor(self, monitor: MonitorSocket, events: PollEventFlag, slot: int) -> None:
        """Register a monitor with ``POLLIN`` and a caller slot.

        Alias of ``add_socket``. Other readiness flags raise ``ConfigError``
        with ``ConfigResult.INVALID_ARGUMENT``. After readiness, drain
        ``monitor.recv(flags=RecvFlags.DONT_WAIT)`` until it returns ``None``.
        Remove the monitor before closing it.
        """
        ...

    def modify_monitor(self, monitor: MonitorSocket, events: PollEventFlag) -> None:
        """Alias of ``modify_socket``; monitors accept only ``POLLIN``."""
        ...

    def remove_monitor(self, monitor: MonitorSocket) -> None:
        """Alias of ``remove_socket``; unregister the borrowed monitor."""
        ...

    def add_fd(self, fd, events, slot):
        """Register raw file descriptor ``fd`` to be watched for ``events``;
        ``slot`` is echoed back in the matching result."""
        ...

    def add_timer(self, timer, slot):
        """Register ``timer``; its expirations surface as poll events tagged
        with ``slot``."""
        ...

    def modify_socket(self, socket, events):
        """Change watched events for a socket or monitor; monitors accept only ``POLLIN``."""
        ...

    def modify_fd(self, fd, events):
        """Change the watched events for an already-registered file
        descriptor."""
        ...

    def remove_socket(self, socket):
        """Unregister a socket or monitor."""
        ...

    def remove_fd(self, fd):
        """Unregister file descriptor ``fd``."""
        ...

    def remove_timer(self, timer):
        """Unregister ``timer``."""
        ...

    def size(self):
        """Return the number of registered sources."""
        ...

    def wait(self, events, timeout_ms):
        """Wait up to ``timeout_ms`` milliseconds for sources to become ready,
        filling ``events``; a negative timeout blocks indefinitely. Return the
        number of ready sources."""
        ...

    def close(self):
        """Close the poller and release its resources."""
        ...

    def __enter__(self): ...

    def __exit__(self, exc_type, exc, tb): ...

    async def __aenter__(self): ...

    async def __aexit__(self, exc_type, exc, tb): ...
